client.go

```diff
@@ -0,0 +1,155 @@
+		if err != nil {
+			panic(err)
+		}
+
+		patient, err := c.Patient.Get("9000000009")
+	}
+
+*/
+
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+type Client struct {
+	BaseURL   *url.URL
+	UserAgent string
+
+	httpClient *http.Client
+
+	common  service // Reuse a single struct instead of allocating one for each service on the heap.
+	Patient *PatientService
+}
+
+//go:generate moq -out client_moq.go . IClient
+type IClient interface {
+	NewRequest(method, path string, body interface{}) (*http.Request, error)
+	Do(ctx context.Context, req *http.Request, v interface{}) (*Response, error)
+}
+
+var errNonNilContext = errors.New("context must be non-nil")
+
+const (
+	defaultBaseURL = "https://sandbox.api.service.nhs.uk/personal-demographics/FHIR/R4/"
+)
+
+// NewClient returns a new FHIR client. If a nil httpClient is provided then a new http.client will be used.
+// To use API methods requiring auth then provide a http.Client which will perform the authentication for you e.g. oauth2
+func NewClient(httpClient *http.Client) *Client {
+	if httpClient == nil {
+		netTransport := &http.Transport{
+			Dial: (&net.Dialer{
+				Timeout: 5 * time.Second,
+			}).Dial,
+			TLSHandshakeTimeout: 5 * time.Second,
+		}
+		httpClient = &http.Client{
+			Timeout:   time.Second * 10,
+			Transport: netTransport,
+		}
+	}
+	baseURL, _ := url.Parse(defaultBaseURL)
+
+	// adds trailing slash
+	if !strings.HasSuffix(baseURL.Path, "/") {
+		baseURL.Path += "/"
+	}
+
+	c := &Client{
+		BaseURL:    baseURL,
+		httpClient: httpClient,
+	}
+
+	c.common.client = c
+
+	// repeat this pattern per service...
+	c.Patient = (*PatientService)(&c.common)
+
+	return c
+}
+
+// NewRequest creates an API request. A relative URL can be provided in path,
+// in which case it is resolved relative to the BaseURL of the Client.
+// Relative URLs should always be specified without a preceding slash. If
+// specified, the value pointed to by body is JSON encoded and included as the
+// request body.
+func (c *Client) NewRequest(method, path string, body interface{}) (*http.Request, error) {
+	rel := &url.URL{Path: path}
+	u := c.BaseURL.ResolveReference(rel)
+	var buf io.ReadWriter
+	if body != nil {
+		buf = new(bytes.Buffer)
+		err := json.NewEncoder(buf).Encode(body)
+		if err != nil {
+			return nil, err
+		}
+	}
+	req, err := http.NewRequest(method, u.String(), buf)
+	if err != nil {
+		return nil, err
+	}
+	if body != nil {
+		req.Header.Set("Content-Type", "application/json")
+	}
+	req.Header.Set("Accept", "application/json")
+	req.Header.Set("User-Agent", c.UserAgent)
+	return req, nil
+}
+
+// Do sends an API request and returns the API response. The API response is
+// JSON decoded and stored in the value pointed to by v, or returned as an
+// error if an API error has occurred.
+func (c *Client) Do(ctx context.Context, req *http.Request, v interface{}) (*Response, error) {
```

Comment:
can this be private?

Reply:
Yep same as NewRequest, will make it private.